jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp

```diff
@@ -0,0 +1,1815 @@
+    {
+      token_p->current_p = current_p + 1;
+      ecma_builtin_json_parse_string (token_p);
+      return;
+    }
+    case 'n':
+    {
+      if (ecma_builtin_json_check_id (current_p, "null"))
+      {
+        token_p->type = null_token;
+        token_p->current_p = current_p + 4;
+        return;
+      }
+      break;
+    }
+    case 't':
+    {
+      if (ecma_builtin_json_check_id (current_p, "true"))
+      {
+        token_p->type = true_token;
+        token_p->current_p = current_p + 4;
+        return;
+      }
+      break;
+    }
+    case 'f':
+    {
+      if (ecma_builtin_json_check_id (current_p, "false"))
+      {
+        token_p->type = false_token;
+        token_p->current_p = current_p + 5;
+        return;
+      }
+      break;
+    }
+    default:
+    {
+      if (*current_p == '-' || isdigit (*current_p))
+      {
+        token_p->current_p = current_p;
+        ecma_builtin_json_parse_number (token_p);
+        return;
+      }
+      break;
+    }
+  }
+} /* ecma_builtin_json_parse_next_token */
+
+/**
+ * Checks whether the next token is right square token.
+ *
+ * @return true if it is.
+ *
+ * Always skips white spaces regardless the next token.
+ */
+static bool
+ecma_builtin_json_check_right_square_token (ecma_json_token_t *token_p) /**< token argument */
+{
+  lit_utf8_byte_t *current_p = token_p->current_p;
+
+  /*
+   * No need for end check since the string is zero terminated.
+   */
+  while (*current_p == ' ' || *current_p == '\r'
+         || *current_p == '\n' || *current_p == '\t')
+  {
+    current_p++;
+  }
+
+  token_p->current_p = current_p;
+
+  if (*current_p == ']')
+  {
+    token_p->current_p = current_p + 1;
+    return true;
+  }
+  return false;
+} /* ecma_builtin_json_check_right_square_token */
+
+/**
+ * Utility for defining properties.
+ *
+ * It silently ignores all errors.
+ */
+static void
+ecma_builtin_json_define_value_property (ecma_object_t *obj_p, /**< this object */
+                                         ecma_string_t *property_name_p, /**< property name */
+                                         ecma_value_t value) /**< value */
+{
+  ecma_property_descriptor_t prop_desc = ecma_make_empty_property_descriptor ();
+  {
+    prop_desc.is_value_defined = true;
+    prop_desc.value = value;
+
+    prop_desc.is_writable_defined = true;
+    prop_desc.is_writable = true;
+
+    prop_desc.is_enumerable_defined = true;
+    prop_desc.is_enumerable = true;
+
+    prop_desc.is_configurable_defined = true;
+    prop_desc.is_configurable = true;
+  }
+
+  ecma_completion_value_t completion_value = ecma_op_object_define_own_property (obj_p,
+                                                                                 property_name_p,
+                                                                                 &prop_desc,
+                                                                                 false);
+  ecma_free_completion_value (completion_value);
```

Comment:
In case, `is_throw` argument of `ecma_op_object_define_own_property` is false, only possible completion values are 'normal true' and 'normal false'. So, we don't need to free the value, but instead can add the following assertion:
```
JERRY_ASSERT (ecma_is_completion_value_normal_true (completion_value)
              || ecma_is_completion_value_normal_false (completion_value));
```